Use safe cast to enable enableEdgeToEdge

atorresveiga opened the pull request:
### Description
This small PR fixes a crash I detected when enabling edge-to-edge on non-ComponentActivities. The fix added here is to use Kotlin's safe cast to check if the activity is a ComponentActivity.

### Testing information
I could reproduce the crash when Leak Canary tried to analyze the app's memory on the app's login flow.

Steps to reproduce
1. Log out from the app
2. Try to log in again until leak canary is triggered
3. Check that the app crashes

### The tests that have been performed
- Checking that the app does not crash when Leak Canary is triggered.

### Images/gif
Crash log:
<img width="700" src="https://github.com/user-attachments/assets/a472db2a-19c7-4ade-ab92-fe20f77b2e78" />


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

Commit:
use safe cast to enable enableEdgeToEdge

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/ApplicationEdgeToEdgeEnabler.kt:
```diff
@@ -9,7 +9,7 @@ import javax.inject.Inject
 class ApplicationEdgeToEdgeEnabler @Inject constructor() : Application.ActivityLifecycleCallbacks {
     override fun onActivityCreated(activity: Activity, bindle: Bundle?) {
-        (activity as ComponentActivity).enableEdgeToEdge()
+        (activity as? ComponentActivity)?.enableEdgeToEdge()
```
I think it's kinda dangerous to do this way. We just won't know if some activity didn't apply edge to edge which can end up in different behavior on 35 and below

I am wondering if we should force edge-to-edge only for activities with our package and do not do anything for the rest?

atorresveiga replied:
> We just won't know if some activity didn't apply edge to edge, which can end up in different behavior on 35 and below

Good point @kidinov.

I refactored the code to:

1. Enable edge-to-edge on supported activities (we have a list of unsupported edge-to-edge activities where we can add exceptional cases)
2. Fail fast when on debug mode, so we can be aware of activities not supporting edge-to-edge
3. Log activities that do not support edge-to-edge when the app is not in debug mode, allowing to continue using the app if this case is reached.

Let me know WDYT?

kidinov reviewed (commented):
Please take a look at the comment I've just left

Commit:
Enable edge-to-edge only for supported activities

Add a check to prevent enabling edge-to-edge on unsupported activities. Report a crash when an activity is not a `ComponentActivity` and is edge-to-edge enabled.

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/ApplicationEdgeToEdgeEnabler.kt:
```diff
@@ -5,11 +5,36 @@ import android.app.Application
+            (activity as ComponentActivity).enableEdgeToEdge()
+        } else {
+            (activity as? ComponentActivity)?.enableEdgeToEdge() ?: run {
+                if (isEdgeToEdgeSupported) {
```
If the activity should support edge-to-edge and it fails, we will log the error without crashing the app

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/ApplicationEdgeToEdgeEnabler.kt:
```diff
@@ -5,11 +5,36 @@ import android.app.Application
+) : Application.ActivityLifecycleCallbacks {
+
+    private val unsupportedEdgeToEdgeActivities = setOf(
+        "leakcanary.internal.RequestPermissionActivity"
```
We can update this set for all the activities that are not `ComponentActivity` and don't need to enable `enableEdgeToEdge()`

kidinov reviewed (approved):
Hey @atorresveiga 

Thanks for the improvement!

My 1 cent on this:

The UI can be broken on that activity which is not Component one and not in "black list"; and I think we will hardly notice those errors in Sentry, especially if that activity is rarely used

Also, the code is complex and by itself can create issues and the "blacklist of the activities" require maintains 

I'd really keep it as simple as that:

```
    override fun onActivityCreated(activity: Activity, bindle: Bundle?) {
        if (activity.isFromCurrentPackage()) {
            (activity as ComponentActivity).enableEdgeToEdge()
        }
    }

    private fun Activity.isFromCurrentPackage(): Boolean {
        return this.packageName == applicationContext.packageName
    }
```

If you think your approach makes more sense - feel free to go with that!

atorresveiga commented:
Thanks for the review, @kidinov, and for sharing your thoughts

> The UI can be broken on that activity which is not Component one and not in "black list"; and I think we will hardly notice those errors in Sentry, especially if that activity is rarely used

In case the activity is not a Component activity and not on the blacklist, we will record the issue on Sentry, so we can notice it and fix it in the future. The important part is that we will record the issue, but the app won't crash, so the merchant can continue using the app.

>  I think we will hardly notice those errors in Sentry, especially if that activity is rarely used

Even if it is a rarely used activity, if this activity is launched from a critical flow (like purchasing a label), I feel it is safer to keep track of the issue instead of blocking the flow with a crash.

> Also, the code is complex and by itself can create issues and the "blacklist of the activities" require maintains

Yep, I tried to provide the best approach depending on whether the app is in production or development and whether it is a known issue, which required some checks. I will add documentation to the function so it is easier to understand what it is all about.

> I'd really keep it as simple as that

Thank you for sharing this approach ☝️. However, this won't solve the issue of separating our code from the external code we use. For instance, the activity that initially prompted this PR is from the LeakCanary library, but it utilizes our package name.

Commit:
add documentation

atorresveiga committed:
Merge branch '13270-android-sdk-update-target-sdk-to-35' into fix/non-component-activity-crash-edge-to-edge

kidinov commented:
@atorresveiga 

> In case the activity is not a Component activity and not on the blacklist, we will record the issue on Sentry, so we can notice it and fix it in the future. The important part is that we will record the issue, but the app won't crash, so the merchant can continue using the app.

My point is that it's not realistic from a practical point of view as nobody will notice those messages in sentry as we have tones of then. It will be just an activity with broken UI for quite sometimes or forever =)

> Thank you for sharing this approach ☝️. However, this won't solve the issue of separating our code from the external code we use. For instance, the activity that initially prompted this PR is from the LeakCanary library, but it utilizes our package name.

Yeah, that was just an idea. We can compare strings, e.g, our activity names start with `com.woocommerce.android`, not ours with something else. The idea is to make sure that we generically apply this to all current and future activities that are "ours" 

As I said If you think your approach makes more sense - feel free to go with that!
